File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const input = require("fs").readFileSync(process.platform === "linux" ? "/dev/stdin" : "알고리즘/example.txt").toString().trim().split("\n");
// 행, 열
const [N, M] = input[0].split(" ").map(Number);
const maze = input.slice(1).map((line) => line.trim().split("").map(Number));
// 인접 방향 순서대로 상, 하, 좌, 우로 1칸 이동할 수 있게 함
const move = [[0, 1], [0, -1], [-1, 0], [1, 0]];

// bfs 시작
function bfs(x, y) {
    // 이동 칸 수를 저장할 배열을 만듦
    let distance = Array.from({ length: N }, () => Array(M).fill(0));
    let queue = [[y, x]];
    distance[y][x] = 1;

    while (queue.length !== 0) {
        // 큐의 첫 번째 요소 -> 현재의 위치 [y, x]
        const [now_y, now_x] = queue.shift();

        // 도착 위치에 도달하면 거리 반환
        if (now_y === N - 1 && now_x === M - 1) {
            return distance[now_y][now_x];
        }

        for (let i = 0; i < 4; i++) {
            // 현재 위치에서의 상하좌우 적용해서 다음 위치 탐색함
            let next_y = now_y + move[i][0];
            let next_x = now_x + move[i][1];

            // 다음 위치가 미로 안 벗어나고 값이 1이면 -> 큐에 그 위치 집어 넣고 현재에서 +1 된 거리를 해당하는 거리 배열에 저장
            if (next_x >= 0 && next_x < M && next_y >= 0 && next_y < N && maze[next_y][next_x] === 1 && distance[next_y][next_x] === 0) {
                queue.push([next_y, next_x]);
                distance[next_y][next_x] = distance[now_y][now_x] + 1;
            }
        }
    }
}

const result = bfs(0, 0);
console.log(result);

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())

maze = []
for _ in range(N):
    maze.append(list(map(int, input().strip())))

visited = [[False] * M for _ in range(N)]

# 오른쪽, 아래, 왼쪽, 위 (상하좌우)
move = [[0, 1], [1, 0], [0, -1], [-1, 0]]

# bfs는 최단거리를 보장한다.
def bfs(x, y):
    queue = deque([(x, y)])  # 시작점 큐에 추가
    visited[x][y] = True
    
    while queue:
        x, y = queue.popleft()
        
        for i in range(4):
            new_x, new_y = x + move[i][0], y + move[i][1]
            # 인덱스 범위를 먼저 확인하고 나서 다른 조건을 체크
            if 0 <= new_x < N and 0 <= new_y < M and not visited[new_x][new_y] and maze[new_x][new_y] == 1:
                # 이동할 수 있는 칸만 큐에 추가하고 방문 처리
                queue.append((new_x, new_y))
                visited[new_x][new_y] = True
                maze[new_x][new_y] = maze[x][y] + 1  # 이동 거리 갱신

bfs(0, 0)

# 최종 목적지까지의 최단 거리 출력
print(maze[N-1][M-1])